File: src/house_robber_198/solution_recursion.go
package house_robber_198

import "fmt"

func maxRob(nums []int, idx int, cache []int) int {
	if idx >= len(nums) {
		return 0
	}

	found := cache[idx]
	if found != -1 {
		return found
	}

	result := max(
		nums[idx]+maxRob(nums, idx+2, cache),
		maxRob(nums, idx+1, cache),
	)

	cache[idx] = result
	return result
}

func robRecursion(nums []int) int {
	cache := make([]int, len(nums))
	for i := range len(cache) {
		cache[i] = -1
	}

	return maxRob(nums, 0, cache)
}

func TestRecursion() {
	fmt.Println(robRecursion([]int{1, 2, 3, 1}))
	fmt.Println(robRecursion([]int{2, 7, 9, 3, 1}))
}

File: src/house_robber_198/solution_recursion.rs
use std::cmp::max;
use std::collections::HashMap;

pub struct SolutionRecursion;

impl SolutionRecursion {
    fn max_rob(nums: &[i32], cache: &mut HashMap<usize, i32>, idx: usize) -> i32 {
        if idx >= nums.len() {
            return 0;
        }

        if let Some(&value) = cache.get(&idx) {
            return value;
        }

        let result = max(
            nums[idx] + Self::max_rob(nums, cache, idx + 2),
            Self::max_rob(nums, cache, idx + 1),
        );

        cache.insert(idx, result);
        result
    }

    pub fn rob(nums: Vec<i32>) -> i32 {
        let mut cache = HashMap::new();

        Self::max_rob(&nums, &mut cache, 0)
    }
}

pub fn test() {
    println!("{}", SolutionRecursion::rob(vec![1, 2, 3, 1]));
    println!("{}", SolutionRecursion::rob(vec![2, 7, 9, 3, 1]));
}
